test-harness/tests/test_resolver.py

```diff
@@ -0,0 +1,121 @@
+# Copyright (C) Mesosphere, Inc. See LICENSE file for details.
+
+import requests
+import time
+
+from runner.common import CACHE_FIRST_POLL_DELAY
+from util import GuardedSubprocess
+
+from generic_test_code import (
+    generic_correct_upstream_dest_test,
+    ping_mesos_agent,
+)
+
+
+class TestNginxResolver:
+
+    # For testing purposes we'll use short DNS records TTL
+    SHORT_TTL = 1
+
+    def test_if_mesos_leader_is_reresolved_in_proxy_pass(
+            self, nginx_class, valid_user_header, dns_server_mock):
+        # Respond with default leader instance and TTL=1s
+        dns_server_mock.reply_with_leader_ip('127.0.0.2', ttl=self.SHORT_TTL)
+
+        ar = nginx_class()
+        with GuardedSubprocess(ar):
+            # Mesos prefixed URL should be forwarded to Mesos leader instance
+            url = ar.make_url_from_path('/mesos/master/state-summary')
+
+            # Make sure that request was proxied to MesosEndpoint mock which
+            # returns data that looks like actual cluster state summary
+            r = requests.get(
+                url, allow_redirects=True, headers=valid_user_header)
+            assert r.status_code == 200
+            assert r.json()["cluster"] == "prozlach-qzpz04t"
+
+            dns_server_mock.reply_with_leader_ip('127.0.0.5', ttl=self.SHORT_TTL)
+            # The MesosLeaderDNSServer resolves leader.mesos with TTL=1s so
+            # its enough to wait for 2 seconds here
+            time.sleep(self.SHORT_TTL * 2)
+
+            # Make sure that second request was forwarded to new leader
+            # which is an instance of ReflectingTcpIpEndpoint
+            r = requests.get(
+                url, allow_redirects=True, headers=valid_user_header)
+            assert r.status_code == 200
+            assert r.json()["endpoint_id"] == "http://127.0.0.5:5050"
+
+    def test_if_mesos_leader_ar_instance_is_reresolved_in_proxy_pass(
```

Comment:
What is the difference between this test and the previous one? Maybe it's too early in the morning for me :(